scaffolding-ruby/lib/scaffolding.sh

```diff
@@ -0,0 +1,946 @@
+
+  if [[ -n "$scaffolding_ruby_pkg" ]]; then
+    _ruby_pkg="$scaffolding_ruby_pkg"
+    build_line "Detected Ruby version in Plan, using '$_ruby_pkg'"
+  else
+    lockfile_version="$($_gemfile_parser ruby-version ./Gemfile.lock || true)"
+    if [[ -n "$lockfile_version" ]]; then
+      # TODO fin: Add more robust Gemfile to Habitat package matching
+      case "$lockfile_version" in
+        *)
+          _ruby_pkg="core/ruby/$(
+            echo "$lockfile_version" | cut -d ' ' -f 2)"
+          ;;
+      esac
+      build_line "Detected Ruby version '$lockfile_version' in Gemfile.lock, using '$_ruby_pkg'"
+    else
+      _ruby_pkg="$_default_ruby_pkg"
+      build_line "No Ruby version detected in Plan or Gemfile.lock, using default '$_ruby_pkg'"
+    fi
+  fi
+  pkg_deps=($_ruby_pkg ${pkg_deps[@]})
+  debug "Updating pkg_deps=(${pkg_deps[*]}) from Scaffolding detection"
+}
+
+_detect_sqlite3() {
+  if _has_gem sqlite3; then
+    build_line "Detected 'sqlite3' gem in Gemfile.lock, adding sqlite packages"
+    pkg_deps=(core/sqlite ${pkg_deps[@]})
+    debug "Updating pkg_deps=(${pkg_deps[*]}) from Scaffolding detection"
+  fi
+}
+
+_detect_webpacker() {
+  if _has_gem webpacker; then
+    build_line "Detected 'webpacker' gem in Gemfile.lock, adding yarn packages"
+    pkg_deps=(core/yarn ${pkg_deps[@]})
+    debug "Updating pkg_deps=(${pkg_deps[*]}) from Scaffolding detection"
+  fi
+}
+
+
+
+
+# **Internal** Invokes the `bundle` program using the chosen version of Ruby.
+# This way, we should avoid most Bundler warnings about mismatched Ruby
+# versions specfied in the Gemfile.
+_bundle() {
+  local bundler_prefix
+  bundler_prefix="$(pkg_path_for bundler)"
+
+  env \
+    -u RUBYOPT \
+    -u GEMRC \
+    GEM_HOME="$bundler_prefix" \
+    GEM_PATH="$bundler_prefix" \
+    "$(pkg_path_for "$_ruby_pkg")/bin/ruby" \
+    "$bundler_prefix/bin/bundle.real" ${*:-}
+}
+
+_rake() {
+  case "$_app_type" in
+    rails*)
+      RACK_ENV=production \
+        RAILS_ENV=production \
+        RAILS_GROUP=assets \
+        _bundle exec rake ${*:-}
+      ;;
+    rack)
+      RACK_ENV=production \
+        _bundle exec rake ${*:-}
+      ;;
+    *)
+      _bundle exec rake ${*:-}
+      ;;
+  esac
+}
+
+_bundle_install() {
+  local path
+  path="$1"
+  shift
+
+  _bundle install ${*:-} \
+    --jobs "$(nproc)" \
+    --without development:test \
+    --path "$path" \
+    --shebang="$(pkg_path_for "$_ruby_pkg")/bin/ruby" \
+    --no-clean \
+    --deployment
+}
+
+_compare_gem() {
+  local gem result
+  gem="$1"
+  shift
+
+  result="$($_gemfile_parser compare-gem-version ${*:-} \
+    ./Gemfile.lock "$gem" 2> /dev/null || true)"
+
+  if [[ "$result" == "true" ]]; then
+    return 0
+  else
+    return 1
+  fi
+}
+
+_create_process_bin() {
+  local bin cmd env_sh
+  bin="$1"
+  cmd="$2"
+  env_sh="$pkg_svc_config_path/hab_env.sh"
+
+  build_line "Creating ${bin} process bin"
+  cat <<EOF > "$bin"
+#!$(pkg_path_for busybox-static)/bin/sh
```

Comment:
This might as well be `#!/bin/sh` or `#!/lol/yolo`. Busybox's sh has, let's say, an indeterminate feature set. Why not just `core/bash`?